### PR TITLE
Search local shards directly to find shard to evict

### PR DESCRIPTION
**Pull request by coryMosaicML**
This PR changes the search for coldest shard to avoid looping over remote shards by considering local shards only as possible candidates for eviction.

### CODE REVIEW

**Comment by j316chuck** [comment]
https://github.com/databricks-mosaic/regression-testing/actions/runs/11078601583 

Adding regression runs here^

**Comment by j316chuck** [comment]
You can invoke like this
Screenshot 2024-09-27 at 3 40 49 PM

**Comment by j316chuck** [comment]
You can invoke like this
Screenshot 2024-09-27 at 3 40 49 PM

**Comment by j316chuck** [comment]
You can invoke like this 
Screenshot 2024-09-27 at 3 41 37 PM

**Comment by j316chuck** [comment]
Screenshot 2024-09-30 at 2 45 49 PM

Throughput looks [normal](https://databricks.slack.com/archives/C05T1A4UMT8/p1727483855679929 
), can we merge?

File: streaming/base/dataset.py [hunk 1]
```diff
@@ -1087,31 +1087,15 @@ def _evict_coldest_shard(self) -> None:
...
-                self._shard_access_times[shard_id] = NEVER
-                continue
-
-            # Break on success.
-            break
-
+        states = self._shard_states.numpy()
+        access_times = self._shard_access_times.numpy()
+        # Filter indices to include only local shards
+        indices = np.where(states == 3)[0]
```
Comment (XiaohanZhangCMU) [inline streaming/base/dataset.py:1093]: suggest: == _ShardState.LOCAL

**Review by XiaohanZhangCMU — CHANGES_REQUESTED** [review]
What if the evicted shard is currently used by other processes? Is there a mechanism to prevent that?

**Review by snarayan21 — COMMENTED** [review]
looks nice! agreed with @XiaohanZhangCMU 's comments. I don't think we actually need a mechanism to make sure the local shard isn't used...according to regression tests, this seems fine to me. probabilistically is probably okay.

**Comment by XiaohanZhangCMU** [comment]
@snarayan21 yeah, I'm ok with it. The question I had is not valid, which should have been handled by the get_item retry logic already.

**Comment by coryMosaicML** [comment]
Closing because it is superseded by [795](https://github.com/mosaicml/streaming/pull/795).

**Pull request closed without merging**